src/Chainweb/Pact/PactService.hs

```diff
@@ -456,62 +464,124 @@ execNewBlock
         -- add the failing transaction to the mempool bad list, so it is not
         -- re-selected for mining.
         liftIO $ mpaBadlistTx mpAccess rk
         throwM e
     onTxFailure e = throwM e
+    -}
 
     -- This is intended to mitigate mining attempts during replay.
     -- In theory we shouldn't need to rewind much ever, but values
     -- less than this are failing in PactReplay test.
     newblockRewindLimit = Just 8
 
-    doPreBlock pdbenv = do
+    getBlockTxs :: BlockFill -> PactServiceM cas (Vector ChainwebTransaction)
+    getBlockTxs bfState = do
       cp <- getCheckpointer
       psEnv <- ask
-      psState <- get
       logger <- view psLogger
-      let runDebitGas :: RunGas
-          runDebitGas txs = evalPactServiceM psState psEnv runGas
-            where
-              runGas = attemptBuyGas miner pdbenv txs
-          validate bhi _bha txs = do
+      let validate bhi _bha txs = do
 
             let parentTime = ParentCreationTime $ _blockCreationTime $ _parentHeader parent
             results <- do
                 let v = _chainwebVersion psEnv
                     cid = _chainId psEnv
-                validateChainwebTxs logger v cid cp parentTime bhi txs runDebitGas
+                validateChainwebTxs logger v cid cp parentTime bhi txs return
 
             V.forM results $ \case
                 Right _ -> return True
                 Left _e -> return False
 
-      liftIO $! fmap Discard $!
-        mpaGetBlock mpAccess validate (pHeight + 1) pHash (_parentHeader parent)
+      liftIO $! {- fmap Discard $! -}
+        mpaGetBlock mpAccess bfState validate (pHeight + 1) pHash (_parentHeader parent)
 
-    doNewBlock newTrans pdbenv = do
-        logInfo $ "execNewBlock, about to get call processFork: "
+    doNewBlock pdbenv = do
+        logInfo $ "execNewBlock: "
                 <> " (parent height = " <> sshow pHeight <> ")"
                 <> " (parent hash = " <> sshow pHash <> ")"
 
+        blockGasLimit <- view psBlockGasLimit
+        let initState = BlockFill blockGasLimit mempty 0
+
+        -- Heuristic: limit fetches to count of 1000-gas txs in block.
+        let fetchLimit = fromIntegral $ blockGasLimit `div` 1000
+
+        newTrans <- getBlockTxs initState
+
         -- NEW BLOCK COINBASE: Reject bad coinbase, always use precompilation
-        results <- execTransactions False miner newTrans
+        (Transactions pairs cb) <- execTransactions False miner newTrans
           (EnforceCoinbaseFailure True)
           (CoinbaseUsePrecompiled True)
           pdbenv
 
-        let !pwo = toPayloadWithOutputs miner results
+        (BlockFilling _ successPairs failures) <-
+          refill fetchLimit pdbenv =<< foldM splitResults (BlockFilling initState mempty mempty) pairs
+
+        liftIO $ mpaBadlistTx mpAccess (V.map gasPurchaseFailureHash failures)
+
+        let !pwo = toPayloadWithOutputs miner (Transactions successPairs cb)
```

Comment:
yes `cb` (coinbase tx) is idempotent as it is executed first and only once.

Reply:
note `refill` uses the new `execTransactionsOnly`

Reply:
> Unrelated to this PR: are the outputs actually needed here? I think returning just the `PayloadData` might be sufficient.

It seems so :) https://github.com/kadena-io/chainweb-node/blob/17d69b99d79c56e282b72e3f19222974de784e18/src/Chainweb/Chainweb/MinerResources.hs#L183

also `WebPactExecutionService` `_webPactNewBlock` is never used, since `MinerResources` are chain-local